op-program/client/interop/cnosolidate.go

```diff
@@ -0,0 +1,227 @@
+package interop
+
+import (
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/op-program/client/interop/types"
+	"github.com/ethereum-optimism/optimism/op-program/client/l2"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/backend/cross"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/backend/depset"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor/backend/processors"
+	supervisortypes "github.com/ethereum-optimism/optimism/op-supervisor/supervisor/types"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+func ReceiptsToExecutingMessages(receipts ethtypes.Receipts) ([]*supervisortypes.ExecutingMessage, error) {
+	var execMsgs []*supervisortypes.ExecutingMessage
+	for _, rcpt := range receipts {
+		for _, l := range rcpt.Logs {
+			execMsg, err := processors.DecodeExecutingMessageLog(l)
+			if err != nil {
+				return nil, err
+			}
+			execMsgs = append(execMsgs, execMsg)
+		}
+	}
+	return execMsgs, nil
+}
+
+func Consolidate(deps ConsolidateCheckDeps,
+	oracle l2.Oracle,
+	transitionState *types.TransitionState,
+	superRoot *eth.SuperV1,
+) (eth.Bytes32, error) {
+	var consolidatedChains []eth.ChainIDAndOutput
+
+	for i, chain := range superRoot.Chains {
+		progress := transitionState.PendingProgress[i]
+		receipts := oracle.ReceiptsByBlockHash(progress.BlockHash, chain.ChainID)
+		execMsgs, err := ReceiptsToExecutingMessages(receipts)
+		if err != nil {
+			return eth.Bytes32{}, err
+		}
+		block := oracle.BlockByHash(progress.BlockHash, chain.ChainID)
+		candidate := supervisortypes.BlockSeal{
+			Hash:      progress.BlockHash,
+			Number:    block.NumberU64(),
+			Timestamp: block.Time(),
+		}
+		hazards, err := getHazards(progress.BlockHash, execMsgs)
+		if err != nil {
+			return eth.Bytes32{}, err
+		}
+		if err := checkHazards(deps, &candidate, hazards); err != nil {
+			// TODO: replace with deposit-only block if ErrConflict, ErrCycle, or ErrFuture
+			return eth.Bytes32{}, err
+		}
+		consolidatedChains = append(consolidatedChains, eth.ChainIDAndOutput{
+			ChainID: chain.ChainID,
+			// TODO: when applicable, use the deposit-only output root
+			Output: chain.Output,
+		})
+	}
+	consolidatedSuper := &eth.SuperV1{
+		Timestamp: superRoot.Timestamp,
+		Chains:    consolidatedChains,
+	}
+	return eth.SuperRoot(consolidatedSuper), nil
+}
+
+type ConsolidateCheckDeps interface {
+	cross.UnsafeFrontierCheckDeps
+	cross.CycleCheckDeps
+}
+
+func checkHazards(
+	deps ConsolidateCheckDeps,
+	candidate *supervisortypes.BlockSeal,
+	hazards map[supervisortypes.ChainIndex]supervisortypes.BlockSeal,
+) error {
+	// TODO: for each of these hazard checks, ensure that they're scoped to the super root timestamp of the chain
```

Comment:
Can we add these TODOs to the notion doc tracking the stuff to do.  When I get a chance I'll convert that page to GitHub tickets.